File: leetcode/CppSession/269_AlienDictionary.cpp
//https://leetcode.com/problems/alien-dictionary/
#include<iostream>
#include<set>
#include<queue>
#include<map>
#include<set>
#include<unordered_map>
#include<unordered_set>
#include<vector>
#include<algorithm>
using namespace std;
class Solution {
    public:
        string alienOrder(vector<string>& words) {
            if(words.empty())   return "";
            if(words.size()==1) return words[0];
            //Form directed graph of dependencies
            unordered_map<char, unordered_set<char> > graph;
            int i, j, k;
            for(i=0;i<words.size();i++){
                for(j=0;j<words[i].size();j++){
                    auto it = graph.find(words[i][j]);
                    if(it == graph.end()){
                        graph[words[i][j]] = unordered_set<char>();
                    }
                }
            }
            string prev_word, word;
            unordered_map<char, int> degrees;
            for(i=1;i<words.size(); i++){
                prev_word = words[i-1];
                word = words[i];
                for(j=0;j<min(prev_word.size(), word.size()); j++){
                    if(word[j] == prev_word[j])
                        continue;
                    auto iit = graph[prev_word[j]].find(word[j]);
                    if(iit == graph[prev_word[j]].end()){
                        graph[prev_word[j]].insert(word[j]);
                        degrees[word[j]]++;
                    }
                    break;
                }
                if(j<=word.size()-1 && j>prev_word.size())
                    return "";
            }

            //Find nodes that have exactly no incoming edges
            queue<char> q;
            for(auto it=graph.begin(); it!=graph.end(); it++){
                char c = it->first;
                //cout<<c<<' '<<degrees[c]<<endl;
                if(degrees[c] == 0)
                    q.push(c);
            }
            if(q.empty())
                return "";
            if(q.size() == degrees.size()){
                if(q.size() > 1){
                    if(words[words.size()-1].size() > words[words.size()-2].size())
                        return words[words.size()-1];
                    return "";
                }
            }

            //BFS over graph
            vector<char> result;
            while(!q.empty()){
                char c = q.front();
                q.pop();
                cout<<c<<endl;

                result.push_back(c);

                auto it = graph.find(c);
                if(it!=graph.end()){
                    //Go to its neighbours and remove the edge from this node to its neighbour
                    for(auto i = (it->second).begin(); i != (it->second).end(); i++){
                        degrees[*i]--;
                        cout<<*i<<' '<<degrees[*i]<<endl;
                        if(degrees[*i] == 0){
                            q.push(*i);
                        }
                    }
                }
            }
            if(result.size()!=degrees.size())
                return "";

            return string(result.begin(), result.end());
        }
};

int main(){
    //string w[] = {"wrt", "wrf", "er", "ett", "rftt"};
    //string w[] = {"z","z"};
    //string w[] = {"wrtkj","wrt"};
    string w[] = {"wrt","wrtkj"};
    //string w[] = {"za","zb", "ca", "cb"};
    vector<string> words(w, w+4);

    class Solution s;
    cout<<endl;

    cout<<s.alienOrder(words)<<endl;
}

File: leetcode/Done/269_AlienDictionary.py
#https://leetcode.com/problems/alien-dictionary/
'''Given a list of non-empty words from the dictionary, where words are sorted
lexicographically by the rules of a new language.
Derive the order of letters in this language.

Inp:
[
  "wrt",
  "wrf",
  "er",
  "ett",
  "rftt"
]
Out: 'wertf'

Inp:
[
  "z",
  "x"
]
Out: 'zx'

Inp:
[
  "z",
  "x",
  "z"
]
Out: '' (Invalid ordering)
'''
import collections
class Solution(object):
    def _addEdge(self, char_a, char_b):
        parents, children = self.parents, self.children
        # Add an edge from A->B by adding A as B's parent and B as A's child
        parents[char_b].add(char_a)
        children[char_a].add(char_b)
        return True

    def findOrder(self, parents, children, chars_seen):
        queue = collections.deque()
        res = []
        # For all the characters present in the word list, pick out characters
        # that don't have any parents as seeds for the topo sort
        for char in chars_seen:
            if char not in parents:
                queue.append(char)

        while queue:
            # For each character in the queue, remove the link to each of its children.
            char = queue.popleft()
            for child in children[char]:
                parents[child].remove(char)
                # If the child has no more parents, then add it to the queue
                if not parents[child]:
                    queue.append(child)
            # Add character to result. And remove it from chars_seen
            res.append(char)
            chars_seen.remove(char)
        # If there are still characters_left, then these haven't been processed
        # ever. Which means these characters are involved in a cycle (inconsistency)
        if chars_seen:
            return ''
        return ''.join(res)

    def alienOrder(self, words):
        if not words:
            return ''
        n = len(words)
        if n==1:
            return ''.join(list(set(words[0])))

        # We attempt to build a graph between characters. Character A->B has an
        # edge, if A lexicographically < B

        # parents[char] is a set of all the parents that a character has
        # children[char] is a set of all the characters that are known to be
        # lexicographically greater than char
        # chars_seen is the set of all unique characters seen in the wordlist
        self.parents = collections.defaultdict(set)
        self.children = collections.defaultdict(set)
        chars_seen = set(words[0])

        for i in xrange(n-1):
            prev_word, next_word = words[i], words[i+1]
            # Adding all the characters of next_word to the char_seen set
            chars_seen.update(next_word)

            # Find the first character at which prev_word and next_word differ.
            # Note: ABC < ABC* (i.e given a common prefix, empty suffix appears
            # before suffixes of any length.
            # Hence, if we traverse through all characters of prev_word, we will
            # find a character that differs in next_word. If not, it means that
            # the prefixes are identical, but prev_word has an empty suffix
            for j, prev_char in enumerate(prev_word):
                if prev_char != next_word[j]:
                    # If characters differ, add an edge to the graph
                    self._addEdge(prev_char, next_word[j])
                    break

        # Run a topological sort and return the ordering
        return self.findOrder(self.parents, self.children, chars_seen)

s = Solution()
inp = [
  "wrt",
  "wrf",
  "er",
  "ett",
  "rftt"
]
print s.alienOrder(inp)

inp = [
  "z",
  "x"
]
print s.alienOrder(inp)

inp = [
  "z",
  "x",
  "z"
]
print s.alienOrder(inp)

inp = ["ri","xz","qxf","jhsguaw","dztqrbwbm","dhdqfb","jdv","fcgfsilnb","ooby"]
print s.alienOrder(inp)

inp = ["bsusz","rhn","gfbrwec","kuw","qvpxbexnhx","gnp","laxutz","qzxccww"]
print s.alienOrder(inp)
